File: target_practice_prompts/swift/w3_d3_sorting.swift
/*
*  Target Practice - Sorting
*
*  Problem 1: Quicksort
*
*  Prompt:    Given an unsorted array of integers, return the array
*             sorted using quicksort.
*
*             What are the time and auxilliary space complexity?
*
*  Input:     An unsorted array of integers
*  Output:    A sorted array of integers
*
*  Example:   input = [3,9,1,4,7] , output = [1,3,4,7,9]
*
*/


// Worse Time Complexity:
// Worse Auxiliary Space Complexity:
// Average Time Complexity:
// Average Auxiliary Space Complexity:
func quicksort(_ input: inout [Int]) -> [Int] {
  // YOUR WORK HERE
  return [Int]()
}




////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

func assert(_ count: inout [Int], _ name: String, _ test: () -> Bool) {
  if count.count != 2 {
    count = [0, 0]
  } else {
    count[1] = count[1] + 1
  }

  var pass: String = "false"

  if test() {
    pass = " true"
    count[0] = count[0] + 1
  }
  print(count[1], ")   ", pass, ":", name)
}

func arraysEqual(_ arr1: [Int], _ arr2: [Int]) -> Bool {
  if arr1.count != arr2.count {
    return false
  }
  for i in 0..<arr1.count {
    if arr1[i] != arr2[i] {
      return false
    }
  }
  return true
}

print("Quicksort Sort Tests")
var testCount: [Int] = [0,0]

assert(&testCount, "should sort [3,9,1,4,7]",  {
  var arr = [3,9,1,4,7]
  let test = quicksort(&arr)
  return arraysEqual(test, [1,3,4,7,9])
})

assert(&testCount, "should return empty array for empty input",  {
  var arr = [Int]()
  let test = quicksort(&arr)
  return arraysEqual(test, [])
})

assert(&testCount, "should sort single-element input",  {
  var arr = [10]
  let test = quicksort(&arr)
  return arraysEqual(test, [10])
})

assert(&testCount, "should sort moderate-sized input",  {
  var arr = [Int]()
  for _ in 1..<1000 {
    arr.append(Int.random(in: 0..<1000))
  }
  var arr2 = arr
  let test = quicksort(&arr2)
  return arraysEqual(test, arr.sorted())
})

assert(&testCount, "should sort large input",  {
  var arr = [Int]()
  for _ in 1..<1000000 {
    arr.append(Int.random(in: 0..<1000000))
  }
  var arr2 = arr
  let test = quicksort(&arr2)
  return arraysEqual(test, arr.sorted())
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n");

File: target_practice_solutions/go/w3_d3_sorting.go
/*
 *  Target Practice - Quicksort
 */

package main
import (
  "fmt"
  "strconv"
  "math/rand"
  "time"
  "sort"
)

/*  Prompt:    Given an unsorted array of integers, return the array
 *             sorted using quicksort.
 *
 *
 *  Input:     input {Array}
 *  Output:    {Array}
 *
 *  Example:   [3,9,1,4,7] --> [1,3,4,7,9]
 */

// Worse Time Complexity: O(N^2)
// Worse Auxiliary Space Complexity: O(N)
// Average Time Complexity: O(Nlog(N))
// Average Auxiliary Space Complexity: O(log(N))
func divide(input []int, begin int, end int) {
  mid := begin
  if begin >= end { return }
  for i := begin; i < end; i++ {
    if input[i] < input[end] {
      input[mid], input[i] = input[i], input[mid]
      mid++
    }
  }
  input[mid], input[end] = input[end], input[mid]
  divide(input, begin, mid - 1)
  divide(input, mid + 1, end)
}

func quicksort(input []int) []int {
   divide(input, 0, len(input) - 1)
   return input
}



////////////////////////////////////////////////////////////
///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
////////////////////////////////////////////////////////////

func main() {
  testCount := []int{0,0}
  fmt.Println("Quick Sort Tests")
  runTest(sortTest1, "should sort example input", testCount);
  runTest(sortTest2, "should return empty array for empty input", testCount);
  runTest(sortTest3, "should sort single-element input", testCount);
  runTest(sortTest4, "should sort moderate-sized input", testCount);
  runTest(sortTest5, "should sort large-sized input", testCount);
  printTestsPassed(testCount);
}

func sortTest1() bool {
  return areEqual(quicksort([]int{3, 9, 1, 4, 7}), []int{1, 3, 4, 7, 9})
}

func sortTest2() bool {
  return areEqual(quicksort([]int{}), []int{})
}

func sortTest3() bool {
  return areEqual(quicksort([]int{10}), []int{10})
}

func sortTest4() bool {
  var RAND_MAX int = 1000
  input := make([]int, RAND_MAX)
  solution := make([]int, RAND_MAX)
  var randNum int
  seed := rand.NewSource(time.Now().UnixNano())
  randObj := rand.New(seed)
  for i := 0 ; i < len(input) ; i++ {
    randNum = randObj.Intn(RAND_MAX) // return a number between ​0​ and RAND_MAX
    input[i] =  randNum
    solution[i] =  randNum
  }
  input = quicksort(input)
  sort.Ints(solution)
  return isSorted(input) && areEqual(input, solution)
}

func sortTest5() bool {
  var RAND_MAX int = 1000000
  input := make([]int, RAND_MAX)
  solution := make([]int, RAND_MAX)
  var randNum int
  seed := rand.NewSource(time.Now().UnixNano())
  randObj := rand.New(seed)
  for i := 0 ; i < len(input) ; i++ {
    randNum = randObj.Intn(RAND_MAX) // return a number between ​0​ and RAND_MAX
    input[i] =  randNum
    solution[i] =  randNum
  }
  input = quicksort(input)
  sort.Ints(solution)
  return isSorted(input) && areEqual(input, solution)
}

/*** Utility Functions ***/
func areEqual(list1 []int, list2 []int) bool {
  if(len(list1) != len(list2)) {
    return false
  }
  for i := 0; i < len(list1) ; i++ {
    if(list1[i] != list2[i]) {
      return false
    }
  }
  return true
}

// checks if array is sorted in linear time
func isSorted(input []int) bool{
  for i := 1 ; i < len(input) ; i++ {
      if input[i-1] > input[i] {
          return false;
      }
  }
  return true;
}

func runTest(test func() bool, testName string, testCount []int) {
  testCount[1]++
  var testPassed bool = test()
  if(testPassed) {testCount[0]++}
  var result string = "  " + (strconv.Itoa(testCount[1]) + ")  ") + strconv.FormatBool(testPassed) + " : " + testName
  fmt.Println(result)
}

func printTestsPassed(testCount []int) {
  fmt.Println("PASSED: " + strconv.Itoa(testCount[0]) + " / " + strconv.Itoa(testCount[1]) + "\n\n")
}
